File: ts/src/day02.ts
import { readFileSync } from "fs";

function getInput() {
    const input = readFileSync("./src/inputs/day02.txt", "utf8");
    return input.split("\n").filter(line => line.length > 0);
}

type Counts = { [key: string]: number }

export function part1() {
    const input = getInput();
    let twoLetter = 0;
    let threeLetter = 0;
    input.map(line => {
        let counts: Counts = {};
        line.split("").map(char => {
            if (counts[char] === undefined) {
                counts[char] = 1;
            } else {
                counts[char] += 1;
            }
        });
        let hasTwo = false;
        let hasThree = false;
        for (let char in counts) {
            if (counts[char] == 3) {
                hasThree = true;
            }
            if (counts[char] == 2) {
                hasTwo = true;
            }
        }
        if (hasThree) {
            threeLetter += 1;
        }
        if (hasTwo) {
            twoLetter += 1;
        }
    });
    console.log(twoLetter, threeLetter);
    return twoLetter * threeLetter;
}

export function part2() {
    const input = getInput();
    for (let i = 0; i < input.length; i++) {
        for (let j = i + 1; j < input.length; j++) {
            let original = input[i].split("");
            let compared = input[j].split("");
            let difference = [];
            for (let c = 0; c < original.length; c++) {
                if (original[c] === compared[c]) {
                    difference.push(original[c]);
                }
            }
            if (difference.length === original.length - 1) {
                return difference.join("");
            }
        }
    }
    
}

File: rs/src/day02.rs
use std::collections::HashMap;

const INPUT: &'static str = include_str!("../inputs/day02.txt");

pub fn part1() -> i32 {
    let input = INPUT.to_string();
    let lines = input.split("\n");

    let mut twos = 0i32;
    let mut threes = 0i32;

    for line in lines {
        let mut chars: Vec<char> = line.chars().collect();
        let mut map: HashMap<char, i32> = HashMap::new();

        for c in chars {
            let new_value = map.get(&c).unwrap_or(&0i32) + 1;
            map.insert(c, new_value);
        }

        let mut has_two = false;
        let mut has_three = false;
        for entry in map.iter() {
            if *entry.1 == 2i32 {
                has_two = true;
            }
            if *entry.1 == 3i32 {
                has_three = true;
            }
        }

        if has_two {
            twos += 1;
        }
        if has_three {
            threes += 1;
        }
    }
    twos * threes
}

pub fn part2() -> String {
    let input = INPUT.to_string();
    let lines: Vec<&str> = input.split("\n").collect();

    for i in 0..lines.len() {
        for j in (i+1)..lines.len() {
            let l1: Vec<char> = lines[i].chars().collect();
            let l2: Vec<char> = lines[j].chars().collect();

            let mut common: Vec<char> = Vec::new();
            for k in 0..l1.len() {
                if l1[k] == l2[k] {
                    common.push(l1[k]);
                }
            }
            if common.len() == l1.len() - 1 {
                return common.iter().collect();
            }
        }
    }

    "".to_string()
}

pub fn run() {
    println!("{}", part1());
    println!("{}", part2());
}
